packages/babel-parser/src/parser/statement.js

```diff
@@ -954,59 +989,66 @@ export default class StatementParser extends ExpressionParser {
       this.unexpected(null, "let is disallowed as a lexically bound name");
     }
     decl.id = this.parseBindingAtom();
-    this.checkLVal(decl.id, true, undefined, "variable declaration");
+    this.checkLVal(
+      decl.id,
+      kind === "var" ? BIND_VAR : BIND_LEXICAL,
+      undefined,
+      "variable declaration",
+    );
   }
 
   // Parse a function declaration or literal (depending on the
   // `isStatement` parameter).
 
   parseFunction<T: N.NormalFunction>(
     node: T,
-    isStatement: boolean,
+    statement?: number = FUNC_NO_FLAGS,
     allowExpressionBody?: boolean = false,
     isAsync?: boolean = false,
-    optionalId?: boolean = false,
   ): T {
-    const oldInFunc = this.state.inFunction;
-    const oldInMethod = this.state.inMethod;
-    const oldInAsync = this.state.inAsync;
-    const oldInGenerator = this.state.inGenerator;
-    const oldInClassProperty = this.state.inClassProperty;
-    const oldYieldPos = this.state.yieldPos;
-    const oldAwaitPos = this.state.awaitPos;
-    this.state.inFunction = true;
-    this.state.inMethod = false;
-    this.state.inClassProperty = false;
-    this.state.yieldPos = 0;
-    this.state.awaitPos = 0;
+    const isStatement = statement & FUNC_STATEMENT;
+    const isHangingStatement = statement & FUNC_HANGING_STATEMENT;
 
     this.initFunction(node, isAsync);
 
+    if (this.match(tt.star) && isHangingStatement) {
+      this.unexpected();
```

Comment:
Maybe we can have a better error message here?

Chrome throws these errors:
```
Generators can only be declared at the top level or inside a block.

Async functions can only be declared at the top level or inside a block.
```

Reply:
I updated them and some more error messages